Some modifications from a Ruby rookie

kbni opened the pull request:
I've made some modifications that I think are useful.

- New Entities - Project and Task
- Picklist helper (load the picklist + reference in queries)
- Query builder (sorry; not using Nokogiri) but generates complex queries nicely
- Selecting Entity fields that are not in Entity.attributes with Entity#[]

This is the first thing I have done in Ruby, so be gentle.

trepidity commented on examples/project_info.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+  end
+  
+  proj_start = Time.parse(proj[:startdatetime]).in_time_zone(TIMEZONE)
```
According to the AutoTask API, time is always done in America/New York. Are you sure about this?

kbni replied:
@trepidity kind of yes, and kind of no. I actually should be converting it to my timezone, and then subtracting the UTC offset back off to get "true midnight" in my timezone. 

Considering the following information from an AutoTask employee.

I'm still very new to Ruby (been using it for like, a week?) and this function is simply used to return information for a Dashboard widget, I shall revisit it however to return the correct true midnight value.

> All DateTimes in the API are Eastern time.   This is noted in the Getting Started section of the API docs.
> 
> Also to note that there are fields where you will see that data is returned with the time portion always “0”.  These are to be considered “date only” and do not require conversion say to a local time

trepidity replied:
For me to make it work right, I had to do two things.

1. Set the time zone in the App ```config.time_zone = 'Eastern Time (US & Canada)'```
2. Parse the date like this. ```Time.zone.parse(date)```

but I'm still learning ruby :)

kbni commented:
@trepidity I've just added some date conversion logic `ActiveSupport::Time`. If you setup `AutotaskAPI::Client` with a TZ, now it will return any field with a name containing 'datetime' as the current time-zone.

Only when using `Entity.field_name` (`Entity[:field_name]` will return the text only). DateTimes are first parsed as `America/New_York` and then converted to `client.tz` unless the datetime is an exact midnight time (ending in T00:00:00) this gets parsed as `client.tz` straight away.

This seems to be working okay for me. I don't want to set app-wide timezone settings (so I can target multiple AutoTasks) and I have used `ActiveSupport::Time` as I am led to believe it has better support for DSTs.

kbni commented:
So @trepidity & @scoop - my additions have gotten a little out of hand. Whilst I've added functionality I need, I'm sure I've done so in a horrible manner. I understand completely if this Pull Request gets closed - but any feedback is appreciated.

trepidity commented:
@kbni, for me, I need to run your changes through my tests to know their impact. I'm not convince that they will work. But don't wait on me, I'm not the owner of Scoop's repo :)

kbni commented:
@trepidity I completely understand - I am quite insure your existing stuff will break if you use this version, although I have tried not to make that the case. Do you happen to have any unit tests written for this module? I will be looking to add this (probably just with minitest) but if you have something..

trepidity commented:
I started to have unit tests, I do everything in rspec, but I ran into a problem feeding XML strings into the API class. The only tests I have, are with the app that use AutoTask_api and those still require manual intervention.

If I can figure how to feed XML into the API, then it would be sweet :)

kbni commented:
@trepidity I agree that makes testing it somewhat tricky - I think as a first step, would it be realistic to have expected values included in the unit tests? For example to have it know to expect certain properties for `YOUR_TEST_TICKET` or `YOUR_TEST_ACCOUNT`?

This would mean the unit tests would remain useful for testing future versions of the API endpoint? I think 1.6 is not far..

scoop commented:
@kbni Alex, first of all, thanks for the PR.

I won't have a chance to look at it in depth in the next week, but I noticed it got to a sizeable chunk of changes since its original submission. Would you agree that it might make sense to break it up into more modular pieces that we can individually review, possibly improve, and then merge?

For instance, the examples I think are very valuable and could be merged with little to no modification, while others are have a deeper impact and need, as discussed previously, either test coverage or more extensive reviews.

I can see at least three different modules that might not even build on top of one another: the query builder, expansion of existing models, and the examples. If you see any additional potential for further packaging those 20 commits, go ahead.

Thanks,
Patrick

kbni commented:
I definitely agree breaking it up would have been the logical thing to do, I am a GitHub & Ruby newbie, and it just uh, kept adding them on top of my initial submission and kind of figured I'd just leave it be. 

Unfortunately my necessity to use the library (and the features I have added) may prevent me from doing so in a timely fashion, so I will close this PR and resubmit when I have added unit testing and cleaned up the code (for example there is query builder code in many source files besides query_builder.rb)

The examples I have provided reference some functionality that only the new query code permits (such as calling a LIKE query using `client.Entity['Foo %']` or calling an object by id `client.Entity[some_id.to_i]`. It's necessary to expand on the models for this functionality, as these methods are used by the Entity object to locate relatable objects (for example, calling `project.tasks` will require a Task entity be available and `ticket.time_entries.first.allocation_code.name` requires a AllocationCode entity)

In the mean time, I would appreciate any feedback towards the code quality (as I'm sure it's not stellar) or how I might otherwise go about cleaning it up - but I understand we are all strapped for time. :)

Cheers,
Alex
